akka-http-core/src/main/scala/akka/http/scaladsl/model/HttpEntity.scala

```diff
@@ -507,15 +507,43 @@ object HttpEntity {
       withSizeLimit(SizeLimit.Disabled)
 
     override def transformDataBytes(transformer: Flow[ByteString, ByteString, Any]): HttpEntity.Chunked = {
-      val newData =
-        chunks.map {
-          case Chunk(data, "")    => data
-          case LastChunk("", Nil) => ByteString.empty
-          case _ =>
-            throw new IllegalArgumentException("Chunked.transformDataBytes not allowed for chunks with metadata")
-        } via transformer
-
-      HttpEntity.Chunked.fromData(contentType, newData)
+      import akka.stream.scaladsl.GraphDSL.Implicits._
+
+      val buf = scala.collection.mutable.ArrayBuffer.empty[Throwable]
+
+      val transformChunks: Flow[ChunkStreamPart, ChunkStreamPart, NotUsed] = Flow.fromGraph(GraphDSL.create() { implicit builder: GraphDSL.Builder[NotUsed] =>
+
+        val bcast = builder.add(Broadcast[ChunkStreamPart](2))
+        val concat = builder.add(Concat[ChunkStreamPart](2))
+
+        val newDataF: Flow[ChunkStreamPart, ByteString, Any] = Flow[ChunkStreamPart].map {
+          case Chunk(data, "")  => data
+          case LastChunk("", _) => ByteString.empty
+        }
+          .via(transformer).recover {
+            case t: Throwable =>
+              buf += t
+              ByteString.empty
+          }
+
+        val toChunksF: Flow[ByteString, ChunkStreamPart, Any] = Flow[ByteString].collect {
+          case b: ByteString if b.nonEmpty => Chunk(b)
+        }
+
+        val trailHeaderF: Flow[ChunkStreamPart, ChunkStreamPart, Any] = Flow[ChunkStreamPart].collect {
+          case lc @ LastChunk(_, s) if s.nonEmpty => lc
+        }.recover {
+          case _: Throwable =>
+            // just needs to complete
+            LastChunk
+        }
+
+        bcast ~> newDataF ~> toChunksF ~> concat
+        bcast ~> trailHeaderF ~> concat
+        FlowShape(bcast.in, concat.out)
```

Comment:
That's an interesting solution but I wonder if it really works. It seems prone to deadlocks because the second part of the stream shouldn't generate any demand until the first part is complete.I wonder why it works in the tests.

Reply:
I haven't thought about dedalocks, just treated that as simple bcast concat flow, can you please explain?

Reply:
It seems to work indeed in this case because `concat` adds a buffer in front of the stage that makes sure there is demand for the second branch even if a potential element will not be consumed until the first branch is done. This works here because the second branch will generate a single element at most.